compiler/rustc_typeck/src/check/generator_interior/drop_ranges.rs

```diff
@@ -0,0 +1,578 @@
+                self.visit_expr(if_true);
+                let true_end = self.expr_count;
+
+                self.drop_ranges.add_control_edge(fork, self.expr_count + 1);
+                if let Some(if_false) = if_false {
+                    self.visit_expr(if_false);
+                }
+
+                self.drop_ranges.add_control_edge(true_end, self.expr_count + 1);
+            }
+            ExprKind::Assign(lhs, rhs, _) => {
+                self.visit_expr(lhs);
+                self.visit_expr(rhs);
+
+                reinit = Some(lhs);
+            }
+            ExprKind::Loop(body, ..) => {
+                let loop_begin = self.expr_count + 1;
+                self.visit_block(body);
+                self.drop_ranges.add_control_edge(self.expr_count, loop_begin);
+            }
+            ExprKind::Match(scrutinee, arms, ..) => {
+                self.visit_expr(scrutinee);
+
+                let fork = self.expr_count;
+                let arm_end_ids = arms
+                    .iter()
+                    .map(|hir::Arm { pat, body, guard, .. }| {
+                        self.drop_ranges.add_control_edge(fork, self.expr_count + 1);
+                        self.visit_pat(pat);
+                        match guard {
+                            Some(Guard::If(expr)) => self.visit_expr(expr),
+                            Some(Guard::IfLet(pat, expr)) => {
+                                self.visit_pat(pat);
+                                self.visit_expr(expr);
+                            }
+                            None => (),
+                        }
+                        self.visit_expr(body);
+                        self.expr_count
+                    })
+                    .collect::<Vec<_>>();
+                arm_end_ids.into_iter().for_each(|arm_end| {
+                    self.drop_ranges.add_control_edge(arm_end, self.expr_count + 1)
+                });
+            }
+            ExprKind::Break(hir::Destination { target_id: Ok(target), .. }, ..)
+            | ExprKind::Continue(hir::Destination { target_id: Ok(target), .. }, ..) => {
+                self.drop_ranges.add_control_edge_hir_id(self.expr_count, target);
+            }
+
+            _ => intravisit::walk_expr(self, expr),
+        }
+
+        self.expr_count += 1;
+        self.drop_ranges.add_node_mapping(expr.hir_id, self.expr_count);
+        self.consume_expr(expr);
+        if let Some(expr) = reinit {
+            self.reinit_expr(expr);
+        }
+    }
+
+    fn visit_pat(&mut self, pat: &'tcx hir::Pat<'tcx>) {
+        intravisit::walk_pat(self, pat);
+
+        // Increment expr_count here to match what InteriorVisitor expects.
+        self.expr_count += 1;
+    }
+}
+
+rustc_index::newtype_index! {
+    pub struct PostOrderId {
+        DEBUG_FORMAT = "id({})",
+    }
+}
+
+rustc_index::newtype_index! {
+    pub struct HirIdIndex {
+        DEBUG_FORMAT = "hidx({})",
+    }
+}
+
+pub struct DropRanges {
```

Comment:
As a meta point, there's an awful lot of code in this file. I think i'd prefer to see the various parts pulled into submodules, with this file being reserved for the really high-level functions that knit everything together -- e.g.

* `drop_ranges/record_consume_borrowed.rs`
* `drop_ranges/cfg_build.rs`
* `drop_ranges/cfg_propagate.rs`

etc